ramp and pwm change direction protection

cychitivav opened the pull request:
Tentative solution for #38

cychitivav committed:
speed slow change

cychitivav committed:
fix negative change

cychitivav committed:
change frequency to 160Hz

cychitivav committed:
add signal shutdown detection

cychitivav committed:
add on shutdown callback

jsduenass commented on scripts/motor_driver.py:
```diff
@@ -15,27 +15,43 @@ def __init__(self, pwm_pin: int, dir_pin: int, pi: pigpio.pi):
-        self.pi.set_PWM_dutycycle(self.pwm_pin, speed)
+        self.speed = 0
+
+    def set_speed(self, speed: float):
+        rate = rospy.Rate(204)  # 120hz
+
+        step = 2*self.MAX_SPEED/100 * (1 if speed > self.speed else -1)
+        instant_speed = self.speed
+        while abs(instant_speed - speed) > abs(step):
+            instant_speed += step
+
+            if instant_speed < 0:
+                abs_speed = -instant_speed
+                dir_value = 1
+            else:
+                abs_speed = instant_speed
+                dir_value = 0
+
+            if abs_speed > self.MAX_SPEED:
+                abs_speed = self.MAX_SPEED
+            
+            # Set direction
+            self.pi.write(self.dir_pin, dir_value)
+            # Set duty cycle to speed
+            self.pi.set_PWM_dutycycle(self.pwm_pin, abs_speed)
+
+            rospy.loginfo("Motor speed: " + str(instant_speed))
+
+            rate.sleep()
```
I like the idea of a ramp. But I consider this is a separate feature that could go into its own function.

jsduenass replied:
@cychitivav 
Here is the branch 
https://github.com/cychitivav/pai/tree/fix/pwm-sign-jsduenass
Should I open a separate pull request?

cychitivav replied:
I think it is not necessary to create a new PR. In my opinion, it is better to solve #38 in this issue and I would propose that you suggest changes in this PR.

> Postscript: I have reviewed your branch and I like this approach, but I think it is still necessary to test this function in the lab. Furthermore, when you suggest changes, I will write some comments.

jsduenass replied:
Thee problem is that you need to import np at the beginning so I find it easier to just add the commit instead of add a suggestion.

jsduenass reviewed (commented):
@cychitivav regarding the ramp approach 

Current approach 
- The ramp adds parameters that may not work well when integrating them with the messages from `/cmd_vel` topic. We want to implement the ramp speed behavior in the commands send by the user to this topic. So I am not sure how these two ramps would work together. I am concern about the response time and delays that could become a issue.

Separate function 
- We can separate function we can switch between them more easily when testing the integration. The downside is that we create two solutions.

I have implemented this independent function approach and would like to hear from you what you think. I would link to the branch

jsduenass committed:
[FEAT]: separate ramp into independent function
- create change of direction protection feature
- remove ramp behavior from set_speed into a independent function

cychitivav commented:
Yes, I think you are right. The ramp may not be the best solution but it is a first step. I am also concerned about the possible issues that may arise later.
